core/src/preflight/util.rs

```diff
@@ -0,0 +1,539 @@
+    Fr, G1,
+};
+use raiko_lib::{
+    builder::{OptimisticDatabase, RethBlockBuilder},
+    clear_line,
+    consts::ChainSpec,
+    inplace_print,
+    input::{
+        ontake::{BlockProposedV2, CalldataTxList},
+        proposeBlockCall, BlobProofType, BlockProposed, BlockProposedFork, TaikoGuestInput,
+        TaikoProverData,
+    },
+    primitives::eip4844::{self, commitment_to_version_hash, KZG_SETTINGS},
+};
+use reth_evm_ethereum::taiko::{decode_anchor, decode_anchor_ontake};
+use reth_primitives::Block;
+use reth_revm::primitives::SpecId;
+use serde::{Deserialize, Serialize};
+use tracing::{debug, error, info, warn};
+
+use crate::{
+    interfaces::{RaikoError, RaikoResult},
+    provider::{db::ProviderDb, rpc::RpcBlockDataProvider, BlockDataProvider},
+    require,
+};
+
+/// Optimize data gathering by executing the transactions multiple times so data can be requested in batches
+pub async fn execute_txs<BDP>(builder: &mut RethBlockBuilder<ProviderDb<BDP>>) -> RaikoResult<()>
+where
+    BDP: BlockDataProvider,
+{
+    let max_iterations = 100;
+    for num_iterations in 0.. {
+        inplace_print(&format!("Executing iteration {num_iterations}..."));
+
+        let Some(db) = builder.db.as_mut() else {
+            return Err(RaikoError::Preflight("No db in builder".to_owned()));
+        };
+        db.optimistic = num_iterations + 1 < max_iterations;
+
+        builder
+            .execute_transactions(num_iterations + 1 < max_iterations)
+            .map_err(|e| {
+                RaikoError::Preflight(format!("Executing transactions in builder failed: {e}"))
+            })?;
+
+        let Some(db) = builder.db.as_mut() else {
+            return Err(RaikoError::Preflight("No db in builder".to_owned()));
+        };
+        if db.fetch_data().await {
+            clear_line();
+            info!("State data fetched in {num_iterations} iterations");
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+/// Prepare the input for a Taiko chain
+pub async fn prepare_taiko_chain_input(
+    l1_chain_spec: &ChainSpec,
+    taiko_chain_spec: &ChainSpec,
+    block_number: u64,
+    l1_inclusion_block_number: Option<u64>,
+    block: &Block,
+    prover_data: TaikoProverData,
+    blob_proof_type: BlobProofType,
+) -> RaikoResult<TaikoGuestInput> {
+    // Decode the anchor tx to find out which L1 blocks we need to fetch
+    let anchor_tx = block
+        .body
+        .first()
+        .ok_or_else(|| RaikoError::Preflight("No anchor tx in the block".to_owned()))?;
+
+    let fork = taiko_chain_spec.active_fork(block.number, block.timestamp)?;
+    info!("current taiko chain fork: {fork:?}");
+
+    let (l1_state_block_number, l1_inclusion_block_number) = match fork {
+        SpecId::ONTAKE => {
+            let anchor_call = decode_anchor_ontake(anchor_tx.input())?;
+            (
+                anchor_call._anchorBlockId,
+                l1_inclusion_block_number.unwrap_or(anchor_call._anchorBlockId + 1),
```

Comment:
Might want to support finding the correct L1 block by brute force running over all 64 possible L1 blocks (starting at _anchorBlockId + 1 and going backwards) because kind of a big ask of people to know which L1 block a specific L2 block was included in.

Reply:
this request comes from client side, they know that num once the block is mined, so send in it is straightforward (just my feeling...), I think it's easier than traversal back 64 blocks for less network access, however, this is based on the assumption that most people use our official client. Will see if we need to support traversal back later. I guess some customized clients running outside, but those high skill people should be able to customize raiko accordingly :)

Reply:
I'm also just thinking from a testing perspective where we'd want to test some L2 blocks, but we of course don't know the corresponding L1 block ourselves, so while it's not ideal to run over the L1 blocks, seems still very useful to do when the L1 block inclusion block number was not specified so the block can still easily be proven.

But agree not that big of a deal.

Reply:
OK, then I can add that back with test purpose later, created issue for it. can do after ontake upgrade.